Migrate legacy Bills tab test

Cruikshanks opened the pull request:
We're onto the last folder of legacy tests; `internal/billing`. No surprise there are tests that have little to do with billing and would be better located elsewhere. The first of these was [mark-licence-for-supplementary-billing.spec.js](https://github.com/DEFRA/water-abstraction-acceptance-tests/pull/43).  This migrates `bills-tab.spec.js`.

We assumed this was checking you can see the bills tab and it has stuff much like `view-contacts.cy.js` which was checking you could see a licence's contacts. So, we moved `search.cy.js` and `view-contacts.cy.js` into a `licences/` folder and got ready to add this legacy test to them.

But when we started the migration we quickly realised it was just a simple test of permissions; whether a certain user should be able to see the bills tab.

We chose to leave the licence tests where they are. It made sense once we'd done it. But we've split the existing `bill-runs.spec.js` test into 2 and added them to `internal/user/permissions/` so it is much clearer what it is they are testing.

andrewhick commented on features/page_objects/internal/account/sign_in.rb:
```diff
@@ -0,0 +1,42 @@
+module Pages
+  module Internal
+    module Account
```
The module structure is new to me. Can you talk me through it at some point? However if it works and saves time then I'm happy.

fayaz0211 replied:
The idea behind using modules here was to prevent name clashes If I want to access the methods or variables with same names been used as I have split the folders into external and internal user journeys it gets easy to trace the files.

andrewhick replied:
Sounds good 👍

andrewhick commented on features/page_objects/common/base_page.rb:
```diff
@@ -0,0 +1,17 @@
+  element(:h1_heading, ".govuk-heading-l")
+  element(:continue_button, ".govuk-button")
+
+  def click_continue
```
I'm happy either way, but does defining a function for a simple, single line command add much value?

fayaz0211 replied:
No I will remove it.

andrewhick commented on features/page_objects/internal/account/sign_in.rb:
```diff
@@ -0,0 +1,42 @@
+        element(:error_heading, ".govuk-error-summary__list")
+        element(:submit_button, "button.govuk-button.govuk-button--start")
+        element(:forgotten_password, "#main-content a:nth-child(1)")
+        element(:create_account_link, "br+ a")
```
Ideally I'd prefer avoid selectors based on relative positions, as the page object might move around. Is there an alternative that might be more stable? Such as `"a[href*='create-account']"`

fayaz0211 replied:
I agree. I used the existing sign in class elements as it is but I can change it.

andrewhick commented on features/page_objects/internal/manage/manage_page.rb:
```diff
@@ -0,0 +1,23 @@
+      end
+
+      def create_a_bill_run
+        find("a", text: "Create a bill run").click
```
I use `find_link("Create a bill run").click which is slightly shorter, but happy if this works for you.

andrewhick commented on features/step_definitions/external/account/sign_in_steps.rb:
```diff
@@ -6,3 +6,12 @@
+
+  data = @test_data.internal_users
+  user_name = data[user]["user_name"]
+  sign_in_page.submit_credentials(user_name, "P@55word")
```
Avoid submitting passwords under a pull request, even dummy ones. Can you call this from an environment variable or the config file instead (assuming this is in your .gitignore)?

fayaz0211 replied:
moved the password to config file

andrewhick replied:
Thanks Fayaz, now we've established that this password is not stored in the database, I'm less worried. However I suggest doing the same for the other password in this file in future.

andrewhick commented on features/page_objects/internal/manage/billing/select_the_region_page.rb:
```diff
@@ -0,0 +1,31 @@
+
+        def submit_region_type(args = {})
+          page.all(:css, ".govuk-radios__item", visible: false).each do |s|
+            if s.text.eql? args[:type]
```
Would `==` also work here? It's probably just a preference thing, but I'd find that more readable.

fayaz0211 replied:
Yes it does. I have used more strictest form of equality operator to compare value and the object type. I can change it to ==

.

andrewhick replied:
No that's fine, if it makes the test stronger then stick with `.eql?` - thanks for confirming.

fayaz0211 commented:
Checked the sign_in_steps and the password been moved to config file for both external and internal login.

fayaz0211 commented:
included the missing changes from the PR review

Cruikshanks committed:
Migrate legacy Bills tab test

We're onto the last folder of legacy tests; `internal/billing`. No surprise there are tests that have little to do with billing and would be better located elsewhere. The first of these was [mark-licence-for-supplementary-billing.spec.js](https://github.com/DEFRA/water-abstraction-acceptance-tests/pull/43).  This migrates `bills-tab.spec.js`.

We assumed this was checking you can see the bills tab and it has stuff much like `view-contacts.cy.js` was checking you could see a licence's contacts. So, we moved `search.cy.js` and `view-contacts.cy.js` into a `licences/` folder and got ready to add this legacy tests to them.

But when we started the migration we quickly realised it was just a simple test of permissions; whether a certain user should be able to see the bills tab.

We chose to leave the licence tests where they are. It made sense once we'd done it. But we've split the existing `bill-runs.spec.js` test into 2 and added them to `internal/user/permissions/` so it is much clearer what it is they are testing.
